tls/s2n_x509_validator.c

```diff
@@ -689,29 +690,50 @@ S2N_RESULT s2n_x509_validator_validate_cert_stapled_ocsp_response(struct s2n_x50
     OCSP_CERTID *cert_id = OCSP_cert_to_id(EVP_sha1(), subject, issuer);
     RESULT_ENSURE_REF(cert_id);
 
+    /**
+     * It is fine to use ASN1_TIME functions with ASN1_GENERALIZEDTIME structures
+     * From openssl documnetation:
+     * It is recommended that functions starting with ASN1_TIME be used instead
+     * of those starting with ASN1_UTCTIME or ASN1_GENERALIZEDTIME. The
+     * functions starting with ASN1_UTCTIME and ASN1_GENERALIZEDTIME act only on
+     * that specific time format. The functions starting with ASN1_TIME will
+     * operate on either format.
+     * https://www.openssl.org/docs/man1.1.1/man3/ASN1_TIME_to_generalizedtime.html
+     */
     ASN1_GENERALIZEDTIME *revtime, *thisupd, *nextupd;
     /* Actual verification of the response */
     const int ocsp_resp_find_status_res = OCSP_resp_find_status(basic_response, cert_id, &status, &reason, &revtime, &thisupd, &nextupd);
     OCSP_CERTID_free(cert_id);
     RESULT_GUARD_OSSL(ocsp_resp_find_status_res, S2N_ERR_CERT_UNTRUSTED);
 
-    uint64_t this_update = 0;
-    RESULT_GUARD(s2n_asn1_time_to_nano_since_epoch_ticks((const char *) thisupd->data,
-            (uint32_t) thisupd->length, &this_update));
+    uint64_t current_sys_time_nanoseconds = 0;
+    RESULT_GUARD(s2n_config_wall_clock(conn->config, &current_sys_time_nanoseconds));
+    /* convert the current_sys_time (which is in nanoseconds) to seconds */
+    time_t current_sys_time_seconds = (time_t) (current_sys_time_nanoseconds / ONE_SEC_IN_NANOS);
+
+    DEFER_CLEANUP(ASN1_GENERALIZEDTIME *current_sys_time = ASN1_GENERALIZEDTIME_set(0, current_sys_time_seconds), s2n_openssl_asn1_time_free_pointer);
 
-    uint64_t next_update = 0;
+    int pday = 0;
+    int psec = 0;
+    /* ensure that current_time is after or the same as "this update" */
+    ASN1_TIME_diff(&pday, &psec, thisupd, current_sys_time);
+    RESULT_ENSURE(pday >= 0 && psec >= 0, S2N_ERR_CERT_INVALID);
+
+    /* ensure that current_time is before or the same as "next update" */
     if (nextupd) {
-        RESULT_GUARD(s2n_asn1_time_to_nano_since_epoch_ticks((const char *) nextupd->data,
-                (uint32_t) nextupd->length, &next_update));
+        ASN1_TIME_diff(&pday, &psec, current_sys_time, nextupd);
+        RESULT_ENSURE(pday >= 0 && psec >= 0, S2N_ERR_CERT_EXPIRED);
```

Comment:
Should this check be OR, not AND?

From the docs:
> ASN1_TIME_diff() sets *pday and *psec to the time difference between from and to. If to represents a time later than from then one or both (depending on the time difference) of *pday and *psec will be positive.

Is there a test that covers this?

Reply:
I think the full documentation specifies this
> ASN1_TIME_diff() sets *pday and *psec to the time difference between from and to. If to represents a time later than from then one or both (depending on the time difference) of *pday and *psec will be positive. If to represents a time earlier than from then one or both of *pday and *psec will be negative. If to and from represent the same time then *pday and *psec will both be zero. **If both pday and psec are non-zero they will always have the same sign.** The value of *psec will always be less than the number of seconds in a day. If from or to is NULL the current time is used.
> --https://www.openssl.org/docs/man1.0.2/man3/ASN1_TIME_diff.html

If current_sys_time is the same as nextup, then the times are the same, and from the documentation both pday and psec will be 0.

If `to:nextupd` represents a time later than `from:current_sys_time 
> then one or both (depending on the time difference) of *pday and *psec will be positive

So possible outcomes are
- outcome 1: ? is positive, ?' is negative
- outcome 2: ? is positive, ?' is zero

but since we know that 
> If both *pday and *psec are non-zero they will always have the same sign.

Then we know that outcome 1 is not possible. So I think the conditions are correct? I agree that it's a little hairy, so possible that I missed something.


> Is there a test that covers this?

I guess, not specifically? I'm not sure what specifically we'd be targetting, but could do something like 
- more than day difference positive, 
- more than day difference negative, 
- less than day different positive, 
- less than day difference negative.

Reply:
Those do sound like good tests if they don't already exist.

Are there any alternatives to this API? Seems like handling its output is potentially a sharp edge, and this really isn't something we want to mess up.